### PR TITLE
[vpj][server][da-vinci][controller] Allow RMD retrieval and parsing in TTL filter at Mapper stage

### PR DESCRIPTION
**Pull request by adamxchen**
## Description

This commit introduced some refactoring changes and the implementation of for RMD retrieval and parsing in TTL filter.

#### [vpj][server][da-vinci]
When repush TTL function is enabled, the VPJ will make a call onto controller to retrieve all RMD schemas belonging to this store and persist them on a temp folder on HDFS. The folder will be unique for each run and gets cleaned up when the push job is finished.
When ttl config is present, both mapper and reducer will create ttl filter. When ttl filter is inited, the schemas will be fetched to further parse the RMD records. Based on the TTL config, the record will be either ignored or filtered.
##### [controller]
The existing API doesn't have the rmd value schema ID info contained, which is essential to differentiate the RMD schemas so also added it.

### Refactoring:
1. Move some Controller/SSL related methods from VPJ main class to an Utils class.
2. Move some Rmd related methods from da-vinci package to the common page as a new Utils class.

### Revision updated on Nov 2, 2022
this commit address some comments in the PR. [(the diff)](https://github.com/linkedin/venice/pull/49/files/44888e566b8a9e845364d35b73e32988602af3ed..ceb87563a344c8c06496f7252332d95ab605ad41)
1) Keep the ttl policy abstraction but remove unused policies for now
2) Let ttl feature uses store-level rewind time and have a flag for customers to turn on explicitly
3) Support chunking but mapper will filter non-chunked records only.
4) When ttl config is present, both mapper and reducer will create ttl filter.
The difference is mapper will only handle non-chunked records whereas reducer can handle both, though in fact reducer will only chunked records if non-chunked records have been processed by mapper.

## How was this PR tested?

internal CI.
JDK 11 - 6664/6665/6666

Added bunch of new unit tests, and integ tests. (See `TestActiveActiveIngestion - testKIFRepushActiveActiveStore` for repush with ttl)

## Does this PR introduce any user-facing changes?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure to explain your proposed changes and call out the behavior change.

See https://github.com/linkedin/venice/pull/10

### Notes

In terms of deployment order, the controller has to be deployed first prior to this VPJ change

### CODE REVIEW

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/ttl/VeniceKafkaInputTTLFilter.java [hunk 1]
```diff
@@ -3,27 +3,67 @@
...
 
   @Override
   protected boolean apply(final KafkaInputMapperValue kafkaInputMapperValue) {
+    Instant curTime = Instant.now();
     switch (ttlPolicy) {
       case RT_WRITE_ONLY:
-        // TODO implementation to retrieve RMD and parse timestamp information. See VENG-9956
-        return false; // do not drop anything
+        Instant timestamp = Instant.ofEpochMilli(getTimeStampFromRmd(kafkaInputMapperValue));
+        return ChronoUnit.HOURS.between(timestamp, curTime) > ttlInHours;
```
Comment (majisourav99) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/ttl/VeniceKafkaInputTTLFilter.java:42]: convert ttl config to ms and compare, else its too coarse grained.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdRmdSchemaSource.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+
+/**
+ * HDFSRmdSchemaSource caches the RMD schemes for a given store on the HDFS and serve them during the Mapper stage
+ * It will make API calls on Venice controller to retrieve all RMD schemes.
+ * <p>
+ * Borrowed some codes from HDFSSchemaSource. The difference is HDFSRmdSchemaSource is created for a single store.
+ */
+@NotThreadsafe
+public class HDFSRmdRmdSchemaSource implements RmdSchemaSource, AutoCloseable {
```
Comment (majisourav99) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdRmdSchemaSource.java:46]: Can we rename the class with RmdRmd->Rmd.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java [hunk 3]
```diff
@@ -1000,6 +987,14 @@ public void run() {
             pushJobSetting.rewindTimeInSecondsOverride = DEFAULT_RE_PUSH_REWIND_IN_SECONDS_OVERRIDE;
             LOGGER.info("Overriding re-push rewind time in seconds to: {}", DEFAULT_RE_PUSH_REWIND_IN_SECONDS_OVERRIDE);
           }
+          if (pushJobSetting.repushTtlInHours != NOT_SET) {
+            // the schema path will be suffixed by the store name, e.g. /tmp/veniceRmdSchemas/<store_name>
```
Comment (majisourav99) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:991]: Should we cleanup the path once we are done?

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdRmdSchemaSource.java (hunk 2 above)
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdRmdSchemaSource.java:46]: Must be a typo :sweat_smile: Thanks for catching it!

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java (hunk 3 above)
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:991]: My current idea is when fetching all schemas from controller, if the schema has persisted, we will skip this, as one minor optimization, so the cleanup is not needed.

Ref:
[code](https://github.com/linkedin/venice/blob/4e510ce4cb349908b2a0eb8dbfee9825a5dcc673/clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdRmdSchemaSource.java#L107)

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdSchemaSource.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+  public void loadRmdSchemasOnDisk() throws IOException, IllegalStateException {
+    if (this.props == null) {
+      throw new IllegalStateException("The schema source has no VeniceProperties provided.");
+    }
+    // there's a risk of closing ControllerClient if passed via the contractor, as the client may be used somewhere else
+    boolean selfClose = false;
+    if (controllerClient == null) {
+      selfClose = true;
+    }
+    ControllerClient client = initControllerClient();
```
Comment (majisourav99) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdSchemaSource.java:100]: even if controllerClient != null, we will create one here. is the expected? also its better to init client outside this scope.

Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdSchemaSource.java:100]: In the `initControllerClient`, if the `controllerClient` is not null, meaning it was inited outside of this class, we will return this one instead creating a new one.

> also its better to init client outside this scope.

I think it's more preferable and also makes codes cleaner. Will update it in the next revision.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/ttl/VeniceKafkaInputTTLFilter.java [hunk 5]
```diff
@@ -3,27 +3,68 @@
...
+  public void closeResources() {
+    schemaSource.close();
+  }
+
+  private Long getTimeStampFromRmd(final KafkaInputMapperValue input) {
+    if (input.replicationMetadataPayload == null || !input.replicationMetadataPayload.hasRemaining()) {
+      throw new IllegalStateException(
+          String.format("The record at offset %s doesn't contain required RMD field.", input.offset));
+    }
+    int id = input.replicationMetadataVersionId, valueSchemaId = input.schemaId;
```
Comment (majisourav99) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/ttl/VeniceKafkaInputTTLFilter.java:61]: nit: each variable in a different line

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/ttl/VeniceKafkaInputTTLFilter.java [hunk 6]
```diff
@@ -3,27 +3,68 @@
...
     }
   }
+
+  @Override
+  public void closeResources() {
+    schemaSource.close();
+  }
+
+  private Long getTimeStampFromRmd(final KafkaInputMapperValue input) {
+    if (input.replicationMetadataPayload == null || !input.replicationMetadataPayload.hasRemaining()) {
```
Comment (majisourav99) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/ttl/VeniceKafkaInputTTLFilter.java:57]: we should allow such records to pass through instead of throwing exception?

File: internal/venice-client-common/src/main/java/com/linkedin/venice/schema/rmd/RmdUtils.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    if (replicationMetadataRecord == null) {
+      return Collections.singletonList(0L);
+    }
+    Object timestampObject = replicationMetadataRecord.get(TIMESTAMP_FIELD_NAME);
+    RmdTimestampType rmdTimestampType = getRmdTimestampType(timestampObject);
+
+    if (rmdTimestampType == RmdTimestampType.VALUE_LEVEL_TIMESTAMP) {
+      return Collections.singletonList((long) timestampObject);
+    } else {
+      // not supported yet so ignore it
```
Comment (majisourav99) [inline internal/venice-client-common/src/main/java/com/linkedin/venice/schema/rmd/RmdUtils.java:84]: I believe this is outdated. We need to scan the per field ts and extract the time stamp.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java (hunk 3 above)
Comment (majisourav99) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:991]: Not sure if we should trust prior runs's temp dir. It could have been corrupted. Is it expensive to repopulate every time ?

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java [hunk 8]
```diff
@@ -2537,8 +2503,13 @@ protected void setupDefaultJobConf(
       conf.set(KAFKA_INPUT_TOPIC, pushJobSetting.kafkaInputTopic);
       conf.set(KAFKA_INPUT_BROKER_URL, pushJobSetting.kafkaInputBrokerUrl);
       conf.setLong(REPUSH_TTL_IN_HOURS, pushJobSetting.repushTtlInHours);
-      conf.setInt(REPUSH_TTL_POLICY, TTLResolutionPolicy.RT_WRITE_ONLY.getValue()); // only support one policy thus not
-                                                                                    // allow any value passed in.
+      if (pushJobSetting.repushTtlInHours != NOT_SET) {
+        conf.setInt(REPUSH_TTL_POLICY, TTLResolutionPolicy.RT_WRITE_ONLY.getValue()); // only support one policy
```
Comment (gaojieliu) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:2507]: In the meeting, we decided to reduce the scope to only support non-WC stores since for WC stores, only looking at the top-level record timestamp will be a semantic change from the current empty push approach.
Can we put this restriction here?

We could talk to the WC customers to understand the desired behavior later.

**Review by gaojieliu — CHANGES_REQUESTED** [review]

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java (hunk 3 above)
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:991]: > It could have been corrupted.

That's a valid point. One way to workaround it is adding permissions so that only venice can modify these files.
>  Is it expensive to repopulate every time ?

We don't have data on how expensive it is yet as it depends on how RMD schemas a store can have.
I think for the time being, we can add the logic to delete the it after the push job is done and tweak this logic as we go. Will do it at the next revision.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java [hunk 9]
```diff
@@ -1000,6 +987,16 @@ public void run() {
             pushJobSetting.rewindTimeInSecondsOverride = DEFAULT_RE_PUSH_REWIND_IN_SECONDS_OVERRIDE;
             LOGGER.info("Overriding re-push rewind time in seconds to: {}", DEFAULT_RE_PUSH_REWIND_IN_SECONDS_OVERRIDE);
           }
+          if (pushJobSetting.repushTtlInHours != NOT_SET) {
+            // the schema path will be suffixed by the store name, e.g. /tmp/veniceRmdSchemas/<store_name>
+            try (HDFSRmdSchemaSource rmdSchemaSource = new HDFSRmdSchemaSource(
+                TEMP_DIR_PREFIX + pushJobSetting.storeName,
```
[the remark this answers is not in the document]
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:993]: Ack.
Will add the logic to clean up old directory such that we always start from fresh tmp folder.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/VeniceKafkaInputMapper.java [hunk 10]
```diff
@@ -54,7 +59,7 @@ protected boolean process(
       BytesWritable keyBW,
       BytesWritable valueBW,
       Reporter reporter) {
-    if (veniceFilter.isPresent() && veniceFilter.get().applyRecursively(inputValue)) {
+    if (veniceFilter != null && veniceFilter.applyRecursively(inputValue)) {
```
[the remark this answers is not in the document]
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/VeniceKafkaInputMapper.java:62]: Yes, you are right! We have this JIRA task in the queue. I intentionally keep that part out of this PR to make this PR size is reasonable.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdSchemaSource.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+  public HDFSRmdSchemaSource(final String schemaDir, final String storeName, final ControllerClient controllerClient)
+      throws IOException {
+    Configuration conf = new Configuration();
+    this.fs = FileSystem.get(conf);
+    this.schemaDir = new Path(schemaDir);
+    if (!fs.exists(this.schemaDir)) {
+      fs.mkdirs(this.schemaDir);
+    }
+    this.storeName = storeName;
+    this.controllerClient = controllerClient;
```
[the remark this answers is not in the document]
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdSchemaSource.java:52]: > Shall we pass a ControllerClient instance to loadRmdSchemasOnDisk

I think that's a good idea, also making the test more easy too!

>  it is good to make loadRmdSchemasOnDisk synchronized to avoid concurrent load and maintain a flag

I thought about that but think `HDFSRmdSchemaSource` is not intended to run in a multi-threading env, in other words, it's just like an Azkaban job, one instance per push job. Also, the load happens on Azkaban job whereas the fetch happens on mapper stage so no flag is needed atm. For the time being, I've added `NotThreadsafe` annotation on this class so people is aware of this. Thanks for this suggestion!

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java (hunk 8 above)
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:2507]: Ack. Will add this!

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/AbstractVeniceFilter.java [hunk 12]
```diff
@@ -58,4 +60,17 @@ public AbstractVeniceFilter<INPUT_VALUE> next() {
...
+
+  public abstract void closeResources() throws IOException;
+
+  /**
+   * Recursively close resources of all filters on the chain. It will start from tail first.
+   * @throws IOException
+   */
+  public void close() throws IOException {
+    if (next != null) {
+      next.close();
```
[the remark this answers is not in the document]
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/AbstractVeniceFilter.java:72]: Hmm.. after thinking it a while, `FilterChain` is more easy-to-understand. Will update it in next revision.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java [hunk 13]
```diff
@@ -1016,8 +1009,24 @@ public void run() {
           pushId = Version.generateRePushId(pushId);
           if (storeSetting.sourceKafkaInputVersionInfo.getHybridStoreConfig() != null
               && pushJobSetting.rewindTimeInSecondsOverride == NOT_SET) {
-            pushJobSetting.rewindTimeInSecondsOverride = DEFAULT_RE_PUSH_REWIND_IN_SECONDS_OVERRIDE;
-            LOGGER.info("Overriding re-push rewind time in seconds to: {}", DEFAULT_RE_PUSH_REWIND_IN_SECONDS_OVERRIDE);
+            // for repush with ttl, default rewind time should be 0 otherwise stale records may be re-consumed.
+            pushJobSetting.rewindTimeInSecondsOverride =
```
Comment (gaojieliu) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:1013]: The TTL is different from rewind time, and even without TTL repush, there is still a rewind time to fill the gap between the real-time write and the new version.

The rewind time should be similar for all the repushes for hybrid stores, but TTL will be very different.

One thing we could consider is to use store-level rewind time as the TTL and version-level rewind time logic should be same as today.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/AbstractVeniceFilter.java [hunk 14]
```diff
@@ -1,61 +1,31 @@
...
-    return this.next;
-  }
-
-  /**
-   * @return true if the chain has more filters.
-   */
-  public boolean hasNext() {
-    return this.next != null;
-  }
+  public abstract void close();
```
Comment (gaojieliu) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/AbstractVeniceFilter.java:30]: `Closeable` already defines a `close` method, why do we need to create a new one here?

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java [hunk 15]
```diff
@@ -329,7 +328,8 @@ public class VenicePushJob implements AutoCloseable {
    */
   public static final String REPUSH_TTL_IN_HOURS = "repush.ttl.hours";
   public static final String REPUSH_TTL_POLICY = "repush.ttl.policy";
```
Comment (gaojieliu) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:330]: I know we supports multiple TTL policies, and I am wondering whether it is necessary to apply TTL to batch data or not, and in theory, TTL repush can only be applied to RT write since for batch-part, we don't know the true write time of the batch data.

I am trying to see how we could simplify TTL repush.

For a real-time only store, I assume TTL repush could be the default mode for repush and no extra config is required, and the TTL will be the rewind time configured in store.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java [hunk 16]
```diff
@@ -1173,6 +1182,9 @@ public void run() {
         pushJobHeartbeatSender.stop();
       }
       inputDataInfoProvider = null;
+      if (pushJobSetting.rmdSchemaDir != null) {
+        HadoopUtils.cleanUpHDFSPath(pushJobSetting.rmdSchemaDir, true);
```
Comment (gaojieliu) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:1186]: If the push job fails to clean up the rmd tmp dir, I assume it will be cleaned up by Grid later automatically, right?

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/VeniceKafkaInputMapper.java [hunk 17]
```diff
@@ -29,22 +30,26 @@ protected AbstractVeniceRecordReader<BytesWritable, KafkaInputMapperValue> getRe
...
   @Override
-  protected Optional<AbstractVeniceFilter<KafkaInputMapperValue>> getFilter(final VeniceProperties props) {
+  protected FilterChain<KafkaInputMapperValue> getFilterChain(final VeniceProperties props) {
+    FilterChain<KafkaInputMapperValue> filterChain = new FilterChain<>();
     long ttlInHours = props.getLong(VenicePushJob.REPUSH_TTL_IN_HOURS, VenicePushJob.NOT_SET);
     if (ttlInHours != VenicePushJob.NOT_SET) {
-      VeniceKafkaInputTTLFilter filter = new VeniceKafkaInputTTLFilter(props);
-      return Optional.of(filter);
+      try {
+        filterChain.add(new VeniceKafkaInputTTLFilter(props));
```
Comment (gaojieliu) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/VeniceKafkaInputMapper.java:38]: Why not use the same constructor to pass a list of filters all the time? And we could remove `add` function from `FilterChain` class.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/VeniceKafkaInputMapper.java [hunk 18]
```diff
@@ -54,7 +59,7 @@ protected boolean process(
       BytesWritable keyBW,
       BytesWritable valueBW,
       Reporter reporter) {
-    if (veniceFilter.isPresent() && veniceFilter.get().applyRecursively(inputValue)) {
+    if (veniceFilterChain != null && !veniceFilterChain.isEmpty() && veniceFilterChain.apply(inputValue)) {
```
Comment (gaojieliu) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/VeniceKafkaInputMapper.java:62]: `veniceFilterChain.isEmpty` check should be handled by `veniceFilterChain.apply`.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdSchemaSource.java [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+   * @throws IOException
+   * @throws IllegalStateException
+   */
+  public void loadRmdSchemasOnDisk(ControllerClient controllerClient) throws IOException, IllegalStateException {
+    LOGGER.info("Starting caching RMD schemas for {} at {}", storeName, schemaDir.toString());
+    MultiSchemaResponse.Schema[] schemas = controllerClient.getAllReplicationMetadataSchemas(storeName).getSchemas();
+    for (MultiSchemaResponse.Schema schema: schemas) {
+      // path for rmd schema is /<schemaDir>/<id>_<valueSchemaId>
+      Path schemaPath = new Path(schemaDir + SEPARATOR + schema.getId() + UNDERSCORE + schema.getRmdValueSchemaId());
+      if (!fs.exists(schemaPath)) {
```
Comment (gaojieliu) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdSchemaSource.java:71]: Shall we throw an exception if the path exists? In theory, the path should be unique since we put the current timestamp part of the file path.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdSchemaSource.java [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+    Configuration conf = new Configuration();
+    this.fs = FileSystem.get(conf);
+    this.schemaDir = new Path(schemaDir);
+    if (!fs.exists(this.schemaDir)) {
+      fs.mkdirs(this.schemaDir);
+    }
+    this.storeName = storeName;
+  }
+
+  public HDFSRmdSchemaSource(final String schemaDirSuffix) throws IOException {
```
Comment (gaojieliu) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdSchemaSource.java:52]: Do we need this constructor?
I assume `storeName` is always known in the map-reduce job.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdSchemaSource.java [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+    for (FileStatus status: fileStatus) {
+      Path path = status.getPath();
+      try (FSDataInputStream in = fs.open(path);
+          BufferedReader reader = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8))) {
+        RmdVersionId pair = parseIdsFromPath(path);
+        String schemeStr = reader.readLine();
+        if (schemeStr == null) {
+          throw new RuntimeException(String.format("Failed to load RMD schema at the path %s", path));
+        }
+        mapping.put(pair, Schema.parse(schemeStr));
```
Comment (gaojieliu) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdSchemaSource.java:105]: Can we use `AvroCompatibilityHelper.parse` instead?

**Review by gaojieliu — CHANGES_REQUESTED** [review]

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/VeniceKafkaInputMapper.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/VeniceKafkaInputMapper.java]: I'll include this change in this PR in my next revision.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java [hunk 22]
```diff
@@ -931,8 +919,13 @@ public void run() {
       storeSetting = getSettingsFromController(controllerClient, pushJobSetting);
       inputStorageQuotaTracker = new InputStorageQuotaTracker(storeSetting.storeStorageQuota);
 
-      if (pushJobSetting.repushTtlInHours != NOT_SET && storeSetting.isChunkingEnabled) {
-        throw new VeniceException("Repush TTL is not supported when the store has chunking enabled.");
+      if (repushWithTTLEnabled(pushJobSetting)) {
+        if (storeSetting.isChunkingEnabled) {
+          throw new VeniceException("Repush TTL is not supported when the store has chunking enabled.");
```
[the remark this answers is not in the document]
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:924]: In the original design, we don't plan to support chunking for TTL repush, at least for now.

The reason is in the reducer has to use composite values to assemble the large value byte array but if they are dropped in the mapper stage due to the TTL, chunking may be broken in this case. We might support it later and it's out of scope at this moment.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java (hunk 13 above)
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:1013]: > One thing we could consider is to use store-level rewind time as the TTL and version-level rewind time logic should be same as today.

IIUC, you are suggesting we use `storeSetting.sourceKafkaOutputVersionInfo.getHybridStoreConfig().getRewindTimeInSeconds()` instead of 0 when customers don't specify the `rewindTimeInSecondsOverride`?
Sure, I think this is more reasonable to gather the information from customer's store setting.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/AbstractVeniceFilter.java (hunk 14 above)
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/AbstractVeniceFilter.java:30]: Good catch! I initially didn't have `Closeable` and forgot to delete this when added it!

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java (hunk 15 above)
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:330]: I know it seems a bit over-engineering but I want to explicitly state that we want to be flexible here in case in the future we are able to know the true write time of the batch data :wink:.

That being said, the `real-time only` is the only policy we support and set as default. Customers won't be able to change it until we support more.

> For a real-time only store, I assume TTL repush could be the default mode for repush and no extra config is required, and the TTL will be the rewind time configured in store.

I am not sure I understand but the TTL repush is not the default mode for real-time only store repush. It will be turn on if and only if `repush.ttl.hours` is specified.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java (hunk 16 above)
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:1186]: I'm not sure about but I think the answer is NO as I've found out people have written bunch of tmp data to `/tmp/`.
We apparently don't want to fail the job due to this but I added one log to track this.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/VeniceKafkaInputMapper.java (hunk 17 above)
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/VeniceKafkaInputMapper.java:38]: Imagining we have multiple filters need to be added to the chain, and we have to add them accordingly inside some `if`/`switch` statement, I think the `add` function will become handy.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdSchemaSource.java (hunk 20 above)
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/schema/HDFSRmdSchemaSource.java:52]: Removed.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java]: Please take a look at my new changes when you have time! Thank you!

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java (hunk 15 above)
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VenicePushJob.java:327]: Have made this feature a flag and use the store-level rewind time

**Review by gaojieliu — CHANGES_REQUESTED** [review]
Looks good overall, and I left some minor comments

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/VeniceKafkaInputReducer.java [hunk 23]
```diff
@@ -60,28 +75,42 @@ protected void setChunkingEnabled(boolean isChunkingEnabled) {
...
-            isEnableWriteCompute(),
-            getDerivedValueSchemaId());
+    if (value == null) {
+      return null;
+    } else if (veniceFilterChain != null && veniceFilterChain.apply(value)) {
+      MRJobCounterHelper.incrRepushTtlFilterCount(reporter, 1L);
+      return null;
+    } else {
+      // TODO: RMD Chunking is not supported, so we don't include RMD info for chunked record.
+      return new VeniceWriterMessage(
```
[the remark this answers is not in the document]
Comment (adamxchen) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/VeniceKafkaInputReducer.java:100]: I meant to say given we haven't supported the RMD chunking in other components, e.g. server, we don't want to include this info in this VeniceWriterMessage just yet.

Will rephrase this to make it more clear.

**Comment by adamxchen** [comment]
@gaojieliu  Thanks for the review! Comments are addressed. Please take a look when you get time!

**Review by gaojieliu — APPROVED** [review]
Thanks a lot!
